Openssl_cert module

MarkusTeufelberger opened the pull request:
##### SUMMARY

Add the openssl_cert module that can generate and validate certificates using various providers.

Currently implemented providers:

- selfsigned (self-sign locally on the remote machine using pyOpenSSL)
- assertonly (does NOT generate any certificate, but only checks an existing certificate - useful to implement your own certificate generation workflow)
- letsencrypt (requests a "Let's Encrypt" certificate using acme-tiny)

See https://github.com/ansible/ansible-modules-extras/pull/2343 for the first commit and initial design/review.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

openssl_cert

##### ADDITIONAL INFORMATION
This builds upon the work by @Spredzy in https://github.com/ansible/ansible-modules-extras/pull/2343

~~This is still a WIP for now (some documentation as well as features are still missing). I'll remove this line as well as adding a comment below once I consider it review-worthy, for now I'm more interested in what the CI bot says.~~

Spredzy committed:
network: Add new module openssl_cert

This module aims to allow a user to manage the lifecycle of SSL
Certificate. Internally it relies on the pyOpenSSL python library
to interact with openssl.

It supports the notion of provided and hence can be enriched with new
providers later on.

Current supported providers are:

  * self-signed: Generate a self-signed OpenSSL certificate
  * letsencrypt: Retrieve a Let's Encrypt OpenSSL certificate

Commit:
Extend openssl_cert module

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,446 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2016, Yanis Guenane <yanis+ansible@guenane.org>
```
Needs updating

MarkusTeufelberger replied:
done

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,446 @@
+
+EXAMPLES = '''
+# Generate a Self Signed OpenSSL certificate
+- openssl_cert: name=www.ansible.com
```
Please use
`key: value` (rather than `key=value`)

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block

MarkusTeufelberger replied:
done

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,446 @@
+
+
+EXAMPLES = '''
+# Generate a Self Signed OpenSSL certificate
```
Please use `-name: ` rather than a comment, as per best practice for writing playbooks

http://docs.ansible.com/ansible/dev_guide/developing_modules_documenting.html#examples-block

MarkusTeufelberger replied:
done

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,446 @@
+        try:
+            cert.generate()
+        except CertificateError:
+            module.fail_json(msg='An error occured while generating the certificate')
```
Could we return any more info than "error?"

MarkusTeufelberger replied:
Not in this state, since this catches any general exceptions from both providers when generating certificates.

MarkusTeufelberger commented:
Thanks a lot already for taking a look, I'll still work on documentation, a proper commit message and a few missing features (checking properties of certificate extensions and validity) tomorrow and will remove the WIP tag once I consider it ready for merging.

If you already have comments about the general interface or architecture, feel free to chime in, for now it is still easy to change stuff.

MarkusTeufelberger commented:
Still WIP, will continue next week.

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,616 @@
+
+notes:
+    - Features/checks that will be skipped with a warning if the version of pyOpenSSL is below a certain version:
+      Check if public key of certificate and public key of private key are equal: requires pyOpenSSL >= 0.16.1
```
The `: ` is confusing the yaml parser, you can remove it or quote the whole line.

gundalow commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,616 @@
+
+    - Features/checks that will be silently skipped if the installed version of cryptography is below a certain version:
+      Expiration checks (except has_expired, which is done with pyOpenSSL): requires cryptography >= 0.7
+      Extensions: requires cryptography >= 1.0
```
The `: ` is confusing the yaml parser, you can remove it or quote the whole line.

MarkusTeufelberger commented:
ready_for_review

If anything is missing or unclear, feel free to comment

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.pycompat24 import get_exception
```
This mechanism has been dropped lately. The correct way of doing it is by following the pattern introduced here https://github.com/ansible/ansible/commit/ed15a0e9a7bf237786634cc5dd89d9448031c86f

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+    def remove(self):
+        '''Remove the certificate'''
+
+        try:
```
One specific case is missing in the  try/catch. See https://github.com/ansible/ansible/commit/ed15a0e9a7bf237786634cc5dd89d9448031c86f

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            self.module.fail_json(msg='The certificate at %s does not exist' % self.dest)
+
+        # pyOpenSSL checks:
+        with open(self.dest, "r") as certfile:
```
Read operation: Ideally should be in a try/catch block, in case file does not exist, permission issue, ...

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+                    msg='Invalid certificate version number (got %s, expected %s)' % (cert_version, self.version))
+
+        if self.privatekey_path is not None:
+            with open(self.privatekey_path, "r") as private_key_file:
```
Read operation: ditto

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+    def load_privatekey(self):
+        '''Load the privatekey object from buffer.'''
+
+        privatekey_content = open(self.privatekey_path).read()
```
Read operation: ditto

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+    def load_csr(self):
+        '''Load the certificate signing request object from buffer.'''
+
+        csr_content = open(self.csr_path).read()
```
Read operation: ditto

Spredzy reviewed (changes requested):
First big thanks for getting this review out of the dust.

This review is partial, I haven't tried it yet, just giving it a first look with first comments. Will go thoroughly later.

1. Exception: The pattern used in the modules have changed lately, I know this commit initially didn't have the new pattern, but this would need to be updated. I posted a link to a review updating it in the other crypto/* modules.

2. Again, mea culpa in my inital review, reads operation are not protected, so nothing guarantee file missing or premission denied. Those would need to be surrounded by a try/catch block

Commit:
Initial round of review

use newer pattern for catching exceptions
try/except for file operations

MarkusTeufelberger commented:
So should I push the changes for this partial review already? I thought it might be better to first wait for the full version, but I of course have already fixed the comments so far locally.

cpointner commented:
bot_status

MarkusTeufelberger commented:
bot_status

@Spredzy @gundalow - please feel free to review! If you want me to push fixes for already addressed issues before you're done, just tell me.

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            dest=dict(required=True, aliases=['path', 'destfile'], type='path'),
```
Replace `dest` by `path` to be consistent with the other modules (openssl_privatekey, openssl_publickey, openssl_csr)

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            provider=dict(required=True, choices=['selfsigned', 'assertonly', 'letsencrypt'], type='str'),
+            force=dict(default=False, type='bool'),
+            backup=dict(default=False, type='bool'),
+            csr=dict(type='str'),
```
Replace `csr` by `csr_path`, also change the type to `path`

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            csr=dict(type='str'),
+
+            # General properties of a certificate
+            privatekey=dict(type='path'),
```
Replace `privatekey` by `privatekey_path`

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            selfsigned_notAfter=dict(default=315360000, type='int'),
+
+            # provider: letsencrypt
+            letsencrypt_accountkey=dict(type='str'),
```
Update type to `path`

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            try:
+                pubkey = privkey.to_cryptography_key().public_key()
+                cert_pubkey = cert.get_pubkey().to_cryptography_key()
+                if pubkey != cert_pubkey:
```
The comparison doesn't work even for valid values.

I've run the following commands:

  * `openssl x509 -noout -modulus -in m.ansible.com.crt | openssl md5`
  * `openssl rsa -noout -modulus -in m.ansible.com.pem | openssl md5`
  * `openssl req -noout -modulus -in m.ansible.com.csr | openssl md5`

They all output the same md5 meaning they're all valid for each other yet this test fails.

I'd guess the content of the object needs to be compared, not the object themselves

MarkusTeufelberger replied:
D'oh! Fixed it by comparing modulus and exponent separately.

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            'notBefore': self.notBefore,
+            'notAfter': self.notAfter,
+            'serial_number': self.serial_number,
+            'backup': backupdest,
```
backupdest isn't defined anywhere

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            'changed': self.changed,
+            'cert_path': self.dest,
+            'account_key': self.accountkey,
+            'backup': backupdest,
```
backupdest isn't defined anywhere

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+        return result
+
+
+class AssertOnlyCertificate(Certificate):
```
provider: assertonly isn't idempotent. It returns change = True when nothing actually changed.

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+    def __init__(self, module):
+        Certificate.__init__(self, module)
+        self.accountkey = module.params['letsencrypt_accountkey']
+        self.challenge_path = module.params['letsencrypt_challenge']
```
module.params['letsencrypt_challenge_path'] is meant here

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+
+        backupdest = ""
+        if self.changed and not self.module.check_mode:
+            if self.backup and os.path.exists(self.dest):
```
Somehow this (+ the following line) doesn't work and prevent the module from working.I haven't digged deeper yet,

MarkusTeufelberger replied:
I got it from https://github.com/ansible/ansible/blob/00ebd0c89d56e5b6c1885d5346bf9f9da0f2f234/lib/ansible/modules/files/lineinfile.py#L398 for example

Still I used it wrongly, since backupdest does not magically become available in the dump() scope, so I fixed that one. :-)

Spredzy commented on lib/ansible/modules/crypto/openssl_cert.py:
```diff
@@ -0,0 +1,996 @@
+            # TODO (spredzy): Ugly part should be done directly by interacting
+            # with the acme protocol through python-acme
+            try:
+                os.system('acme-tiny --account-key %s --csr %s --acme-dir %s > %s' % (self.accountkey, self.csr_path, self.challenge_path, self.dest))
```
Can we remplace this line by the following three lines:

```
 p = subprocess.Popen(['acme-tiny', '--account-key', self.accountkey, '--csr', self.csr_path, '--acme-dir', self.challenge_path], stdout=subprocess.PIPE)
crt = p.communicate()[0]                                                                                                                                                     
open(self.dest, 'w').write(str(crt))
```

Spredzy reviewed (changes requested):
Ok, I could give it a full functional tests.

Conclusion:

Once some of the changed below applied I could run

  * selfsigned
  * assertonly
  * letsencrypt

assertonly is not idempotent, this leads the module to always return change = True, this needs to be changed.

letsencrypt needs few changes (written as comment) to work out of the box.


Also, can we please separate this review in two reviews as the review is a really big change.

Review 1: The module without the check function, simply the module the various providers and that's it

Review 2: Integration of the check function.

It will definitely ease the review. If you have any question you'd like to ask, you can find me on Freenode as spredzy

Commit:
First round of review

letsencrypt_accountkey to type path
privatekey -> privatekey_path
csr -> csr_path
dest -> path
don't call acme-tiny via os.system
fix letsencrypt_challenge_path
compare modulus and exponent of pubkeys separately
fix assertonly idempotency
fix backupdest scope

MarkusTeufelberger commented:
Thanks a lot! Yeah, the size grew a bit larger than expected as one of my internal requirements is that it should also run on Debian Jessie, so it needed a few workarounds to support library versions as old as possible. Sorry for that.

Of course I'm happy to split the review in several parts, I've pushed the fixes for now.

MarkusTeufelberger commented:
Not to self: Write tests once https://github.com/ansible/ansible/pull/26684 lands.

MarkusTeufelberger commented:
I'll add tests in a separate pull request by the way, I still consider this one to be mergeable already.

MarkusTeufelberger commented:
Pinging @Spredzy and @gundalow - I'll be ooo until mid August and still really want to land this for 2.4.

Since the community module freeze will be on August 29th, I really hope that you'll be able to take another look and approve this or tell me where to improve it until then. If the integration test PR lands in the meantime, I'll of course also help writing a few tests for this as well.

bot_status

Spredzy commented:
@MarkusTeufelberger Hey. Last week, the based code has been refactored in order to gain consistence over the various `openssl_` modules.

You can see the examples for:

  * `openssl_privatekey`: https://github.com/ansible/ansible/pull/27221
  * `openssl_publickey`: https://github.com/ansible/ansible/pull/27231

I've notified @cpointner about that and he answered he'll be working on `openssl_csr` updated. (https://github.com/ansible/ansible/pull/26138).

My question is, since you'll be ooo until mid-august:

  * Can I take over this change, refactor it, and have you as Co-Author of the full thing?
  * Would you rather @cpointner (IIRC he's your coworker) to do it?
  * Do you prefer to do the refacto yourself ?

Let us know, and we'll take a decision based on your answer. We all want the same thing here, this to land for 2.4 :)

MarkusTeufelberger commented:
I'd prefer if you refactor it, it seems to be mainly just regarding the check() function anyways.

If you would rather work on your integration tests etc. that's also no problem, I can dedicate some time after I'm back to fix it as well in case you don't want to dive deep into this. @cpointner said he'd prefer to stay out of it and wants to stay focused on csrs.

Spredzy commented:
@MarkusTeufelberger ok, I'll take care of it and cc you on the pull request once made. I'll make sure to also add a comprehensive suite of test now that https://github.com/ansible/ansible/pull/26684 has been merged.

Commit:
Add small integration test

More to come

Commit:
fix test

MarkusTeufelberger commented:
Added a small integration test to see how it works and rebased onto current devel branch.

Looking through the crypto_utils and how you refactored openssl_privatekey:
Do you want me to refactor all the individual checks into private functions similar to what you did in the check() function in openssl_privatekey (_check_size(privatekey), _check_type(privatekey) etc.)?

Commit:
fix pylint errors

Commit:
move from __future__ import up

mattclay commented:
CI failure in integration tests on python 2:

```
Traceback (most recent call last):
  File "/tmp/ansible_sBcqFq/ansible_module_openssl_cert.py", line 1023, in <module>
    main()
  File "/tmp/ansible_sBcqFq/ansible_module_openssl_cert.py", line 1009, in main
    cert.generate()
  File "/tmp/ansible_sBcqFq/ansible_module_openssl_cert.py", line 839, in generate
    cert.add_extensions(self.request.get_extensions())
AttributeError: 'X509Req' object has no attribute 'get_extensions'
```

CI failure in integration tests on python 3:

```
Traceback (most recent call last):
  File "/tmp/ansible_1ulxgfty/ansible_module_openssl_cert.py", line 1023, in <module>
    main()
  File "/tmp/ansible_1ulxgfty/ansible_module_openssl_cert.py", line 1009, in main
    cert.generate()
  File "/tmp/ansible_1ulxgfty/ansible_module_openssl_cert.py", line 851, in generate
    cert_file.write(crypto.dump_certificate(crypto.FILETYPE_PEM, self.certificate))
TypeError: write() argument must be str, not bytes
```

MarkusTeufelberger commented:
The python 2 error is likely because of a too old pyOpenSSL version and could use a better error description (not much more that one can do here), for the python 3 error probably "wb" instead of "w" should be enough. Spredzy is currently refactoring the module, I'll fix these issues when he's done and rather will write a few more tests to uncover more potential issues like this one.

MarkusTeufelberger commented:
See #28263 for the next iteration done by Spredzy.

MarkusTeufelberger commented:
Closing this one in favor of #28263
